### PR TITLE
bal 1379 assign alert bulk operation

### PR DESCRIPTION
**Pull request by liorzam**

### CODE REVIEW

File: services/workflows-service/prisma/schema.prisma [hunk 1]
```diff
@@ -676,6 +676,7 @@ model Alert {
 
   assignee   User?   @relation(fields: [assigneeId], references: [id], onUpdate: Cascade, onDelete: NoAction)
   assigneeId String?
+  assigneeAt DateTime?
```
Comment (alonp99) [inline services/workflows-service/prisma/schema.prisma]: assignedAt*

Comment (alonp99) [reply services/workflows-service/prisma/schema.prisma]: assignedBy may also be vauleable

**Review by alonp99 — APPROVED** [review]

File: services/workflows-service/src/alert/alert.controller.external.ts [hunk 2]
```diff
@@ -6,42 +6,107 @@ import { PrismaService } from '@/prisma/prisma.service';
...
 
-  @Get('/')
+  @common.Get('/')
   @swagger.ApiOkResponse({ type: Array<Object> }) // TODO: Set type
   @swagger.ApiNotFoundResponse({ type: errors.NotFoundException })
   @swagger.ApiForbiddenResponse({ type: errors.ForbiddenException })
-  @UsePipes(new ZodValidationPipe(FindAlertsSchema, 'query'))
-  async getAll(@Query() findAlertsDto: FindAlertsDto, @ProjectIds() projectIds: TProjectId[]) {
+  @common.UsePipes(new ZodValidationPipe(FindAlertsSchema, 'query'))
+  async getAll(
```
Comment (Omri-Levy) [inline services/workflows-service/src/alert/alert.controller.external.ts]: Please follow existing naming conventions, we're open to changing them but in an organized manner. Can be `listAlerts` or similar too.
```suggestion
  async list(
```

File: services/workflows-service/src/alert/alert.repository.ts [hunk 3]
```diff
@@ -42,6 +42,33 @@ export class AlertRepository {
...
   }
 
+  // Method to update an alerts
+  async updateMany<T extends Omit<Prisma.AlertUpdateManyArgs, 'where'>>(
+    alertIds: string[],
+    projectId: TProjectId,
+    args: Prisma.SelectSubset<T, Omit<Prisma.AlertUpdateManyArgs, 'where'>>,
+  ): Promise<Alert[]> {
+    if (!projectId)
+      throw new Error('Project ID is required to perform an update opeartion on Alerts');
```
Comment (Omri-Levy) [inline services/workflows-service/src/alert/alert.repository.ts]: Should it be `Error` and not one of our custom classes?
```suggestion
    if (!projectId) {
      throw new Error('Project ID is required to perform an update opeartion on Alerts');
   };
```

Comment (Omri-Levy) [inline services/workflows-service/src/alert/alert.repository.ts]:
```suggestion
    if (!projectId)
      throw new Error('Project ID is required to perform an update operation on Alerts');
```

File: services/workflows-service/src/alert/alert.service.ts [hunk 4]
```diff
@@ -20,6 +22,32 @@ export class AlertService {
...
+  ): Promise<Alert[]> {
+    try {
+      return await this.alertRepository.updateMany(alertIds, projectId, {
+        data: {
+          assigneeId: assigneeDto.assigneeId,
+        },
+      });
+    } catch (error: unknown) {
+      // Should be handled by ProjectAssigneeGuard on controller level
+      if (error instanceof Prisma.PrismaClientKnownRequestError) {
```
Comment (Omri-Levy) [inline services/workflows-service/src/alert/alert.service.ts]: Replace with a method similar to `isRecordNotFoundError` found in `prisma.util.ts`.

File: services/workflows-service/src/alert/alert.service.ts [hunk 5]
```diff
@@ -20,6 +22,32 @@ export class AlertService {
...
+    projectId: string,
+    assigneeDto: AlertAssigneeUniqueDto,
+  ): Promise<Alert[]> {
+    try {
+      return await this.alertRepository.updateMany(alertIds, projectId, {
+        data: {
+          assigneeId: assigneeDto.assigneeId,
+        },
+      });
+    } catch (error: unknown) {
```
Comment (Omri-Levy) [inline services/workflows-service/src/alert/alert.service.ts]: `unknown` is supposed to be the default.
```suggestion
    } catch (error) {
```

File: services/workflows-service/src/alert/dtos/assign-alert.dto.ts [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+  })
+  status!: string;
+}
+
+export class BulkAssignAlertsResponse {
+  @ApiProperty({
+    required: true,
+    enum: BulkStatus,
+  })
+  overallStatus!: TBulkStatus;
```
Comment (Omri-Levy) [inline services/workflows-service/src/alert/dtos/assign-alert.dto.ts:62]: @alonp99 did we agree that the HTTP code can be used over `overallStatus`?

File: services/workflows-service/src/alert/guards/project-assignee.guard.ts [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+      assigneeId = request.body.assigneeId;
+    } else if (request.query.assigneeId && typeof request.query.assigneeId === 'string') {
+      assigneeId = request.query.assigneeId;
+    } else if (request.params.assigneeId) {
+      assigneeId = request.params.assigneeId;
+    }
+
+    if (!assigneeId) {
+      this.logger.warn('Assignee id not found in request');
+      return false;
```
Comment (Omri-Levy) [inline services/workflows-service/src/alert/guards/project-assignee.guard.ts]: A line break like other places would be appreciated. 🙏🏼

File: services/workflows-service/src/alert/types.ts [hunk 8]
```diff
@@ -5,3 +5,33 @@ export type TAuthenticationConfiguration = {
   isValid: boolean;
   webhookSharedSecret: string;
 };
+
+export const BulkStatus = {
+  success: 'success',
+  partial: 'partial',
+  failed: 'failed',
```
Comment (Omri-Levy) [inline services/workflows-service/src/alert/types.ts]: Let's keep at least the keys consistent with the rest of the system. 🙏🏼 
```suggestion
  SUCCESS: 'success',
  PARTIAL: 'partial',
  FAILED: 'failed',
```

File: services/workflows-service/src/alert/types.ts [hunk 9]
```diff
@@ -5,3 +5,33 @@ export type TAuthenticationConfiguration = {
...
+  alertId: string;
+  status: 'success';
+};
+
+type FailedResponse = {
+  alertId: string;
+  status: 'failed';
+  error: {
+    message: string;
+  };
```
Comment (Omri-Levy) [inline services/workflows-service/src/alert/types.ts]: Don't think errors should be handled differently in different places.
```suggestion
  errors: {
    path?: string;
    message: string;
  }[];
```

File: services/workflows-service/src/case-management/case-management.service.ts [hunk 10]
```diff
@@ -66,7 +66,7 @@ export class CaseManagementService {
     const isValid = validate(entity);
 
     if (!isValid) {
-      throw ValidationError.fromAjvError(validate.errors!);
+      throw BadValidationException.fromAjvError(validate.errors!);
```
Comment (Omri-Levy) [inline services/workflows-service/src/case-management/case-management.service.ts:69]: Is the rename due to the naming conflict with class-validator?

File: services/workflows-service/src/errors.ts [hunk 11]
```diff
@@ -75,6 +76,15 @@ export class ValidationError extends common.BadRequestException {
       path: zodIssue.path.join('.'), // Backwards compatibility - Legacy code message excepts array
     }));
 
-    return new ValidationError(errors);
+    return new BadValidationException(errors);
+  }
+
+  static fromClassValidator(error: ValidationError[]) {
```
Comment (Omri-Levy) [inline services/workflows-service/src/errors.ts:82]: Please consider checking out of dev for a moment and making unrelated changes on a separate PR. 🙏🏼

File: services/workflows-service/prisma/schema.prisma (hunk 1 above)
Comment (liorzam) [reply services/workflows-service/prisma/schema.prisma]: assignedBy currently is out of scope

File: services/workflows-service/src/alert/dtos/assign-alert.dto.ts (hunk 6 above)
Comment (liorzam) [reply services/workflows-service/src/alert/dtos/assign-alert.dto.ts:62]: this schema provided by @alonp99 
let me know what status code would you expect to ?

File: services/workflows-service/src/alert/types.ts (hunk 9 above)
Comment (liorzam) [reply services/workflows-service/src/alert/types.ts]: its not a validation error, this kind of error handling by class-validator before reaching the Controller api

File: services/workflows-service/src/case-management/case-management.service.ts (hunk 10 above)
Comment (liorzam) [reply services/workflows-service/src/case-management/case-management.service.ts:69]: yeah, i could use 'as' which its confusing to use it and rename this class name in some places

File: services/workflows-service/src/errors.ts (hunk 11 above)
Comment (liorzam) [reply services/workflows-service/src/errors.ts:82]: its related to my pr when i couldnt see validation errors, the api didnt work as expected

File: services/workflows-service/src/alert/types.ts (hunk 9 above)
Comment (Omri-Levy) [reply services/workflows-service/src/alert/types.ts]: I understand where you're coming from but as a client it would be unpredictable and require more inside knowledge (even if from Swagger) if sometimes its `error` sometimes `errors` with different data structures.

File: services/workflows-service/src/errors.ts (hunk 11 above)
Comment (Omri-Levy) [reply services/workflows-service/src/errors.ts:82]: You're fixing class validator's errors. Its affecting your PR but its unrelated.

File: services/workflows-service/src/alert/alert.service.ts (hunk 4 above)
Comment (liorzam) [reply services/workflows-service/src/alert/alert.service.ts]: done

File: services/workflows-service/src/alert/dtos/assign-alert.dto.ts (hunk 6 above)
Comment (alonp99) [reply services/workflows-service/src/alert/dtos/assign-alert.dto.ts:62]: We can use only status code or both status code and overallStatus

**Review by alonp99 — APPROVED** [review]

File: services/workflows-service/src/alert/alert.service.ts [hunk 12]
```diff
@@ -20,6 +23,29 @@ export class AlertService {
...
+  ): Promise<Alert[]> {
+    try {
+      return await this.alertRepository.updateMany(alertIds, projectId, {
+        data: {
+          assigneeId: assigneeDto.assigneeId,
+        },
+      });
+    } catch (error) {
+      // Should be handled by ProjectAssigneeGuard on controller level
+      if (error instanceof Prisma.PrismaClientKnownRequestError) {
```
Comment (Omri-Levy) [inline services/workflows-service/src/alert/alert.service.ts]: Looks like this condition is part of `isFkConstraintError`, can we remove this if?

File: services/workflows-service/src/alert/types.ts (hunk 9 above)
Comment (liorzam) [reply services/workflows-service/src/alert/types.ts]: Please resolve and we will discuss about it later with alon

File: services/workflows-service/src/alert/dtos/assign-alert.dto.ts (hunk 6 above)
Comment (liorzam) [reply services/workflows-service/src/alert/dtos/assign-alert.dto.ts:62]: Can you explain what status code would you expect when overallStatus is `partial`?

**Review by Omri-Levy — APPROVED** [review]
